[DOCU-1898] Update custom themes documentation (#30)
Kong/insomnia-docs: docs/insomnia/custom-themes.md

category-url: plugins
---

Make a custom Insomnia theme by creating an Insomnia [plugin](/insomnia/introduction-to-plugins/).

{:.alert .alert-primary}
**Note**: To get started with some live examples, see our built-in [insomnia-plugin-themes](https://github.com/Kong/insomnia/tree/develop/plugins/insomnia-plugin-core-themes) module.

## How to Build a Custom Theme

1. Create a new project for your theme. If you plan to make this theme available via NPM, start your project name with `insomnia-plugin-`. The following example is called `insomnia-plugin-dark-colorblind-theme`.

2. Following plugin instructions, write your [plugin package.json](https://docs.insomnia.rest/insomnia/introduction-to-plugins#plugin-packagejson). In your entry file, export your theme.

3. Start with a baseline template. Each section of the configuration has `background` and `foreground` options, with seven (7) modifiable color options. 
   

>    See the code comments for more information about which color option names correspond with which UI elements.

Reviewer: I'm a bit torn on listing which property corresponds to which UI element, because it also depends on which `styles.*` property the colors are specified on.

The behavior with themes is more that users specify the bg/font colors for `surprise`, `danger`, `notice`, and the `highlight` colors of `xxs, xs, sm, ...` and Insomnia applies those to elements which could always change. This follows the bootstrap pattern to my knowledge.

Also maybe worth looking at VS Code and how they document [theme extensions](https://code.visualstudio.com/api/extension-guides/color-theme) though their [theme API](https://code.visualstudio.com/api/references/theme-color) is much better than ours. 

It's probably okay to give these example in comments but I think it's worth saying that it's not an exhaustive list?

Author: I can add that caveat. However, we need to give developers _some_ ground to stand on.

Reviewer: Yep that's all good 👍🏽 thanks

```ts
module.exports = {
  name: 'Dark Colorblind',        // full theme name

>   displayName: 'Dark Colorblind', // theme name as it will appear on the Insomnia themes tab

Author: It's intentional as it differentiates the two. Yes, we can shorten. However, can you explain more clearly the difference between `name` and `displayName`. I don't have a good idea of a case in which they would be different.

>   displayName: 'Dark Colorblind', // theme name as it will appear on the Insomnia themes tab

Reviewer: Oh interesting, looking at the other themes in `insomnia-plugin-core-themes`, the name is kebab-case. It'd be a unique id, whereas the display name is just the display name.

Worth changing the example here to be `name: 'dark-colorblind'`

>   displayName: 'Dark Colorblind', // theme name as it will appear on the Insomnia themes tab

Reviewer: Oh interesting, looking at the other themes in `insomnia-plugin-core-themes`, the name is kebab-case. It'd be a unique id, whereas the display name is just the display name.

Worth changing the example here to be `name: 'dark-colorblind'`

  theme: {
    // background and foreground nested directly in the theme object will serve as the default overwrites for all options you add
    background: {
      default:    '#21262D',  // primary background color
      success:    '#1F6FEB',  // POST request, 200 OK, parameter names

>       notice:     '#E8F086',  // SEND button, GET request

Author: Does `notice` cover anything else other than the GET request text?

>       notice:     '#E8F086',  // SEND button, GET request

Reviewer: From what I see at a glance it's used for the PATCH (GET is surprise), error notices in various places, markdown previews, and there are probably more.

All of the color schemes (success, notice, etc) and highlights have many uses which is why I think it's important to indicate the list isn't exhaustive

      notice:      '#000',    // secondary font color for notice background
      warning:     '#fff',    // secondary font color for warning background
      danger:      '#fff',    // secondary font color for danger background
      surprise:    '#000',    // secondary font color for surprise background
      info:        '#000'     // secondary font color for info background
    },
    highlight: {
      default: '#D3D3D3'      // sidebar highlight color
    },
    // the styles object targets sub-components of the Insomnia application
    styles: {
      appHeader: {
        foreground: {
          surprise:   '#000'  // header branch button font color
        }
      },
      paneHeader: {
        foreground: {
          surprise:   '#000', // pane accent font color
          info:       '#000'  // pane response font color
        }
      },
      editor: {
        foreground: {
          default:    '#000', // primary editor font color
          surprise:   '#000', // editor accent font color
          info:       '#000'  // editor response font color
        }
      },
      dialog: {
        background: {

>           default:    '#2E4052' // modal primary background color

Author: I always like to front-load the important information.

>           default:    '#2E4052' // modal primary background color

Reviewer: Was just following the pattern from the one below (foreground.default) for consistency. Should that one be changed to `modal primary font color`?

        },
        foreground: {
          default:    '#fff'    // primary font color for modals
        }
      }
    }
  }
}
```

## Styles Parameters

> The following `style` options are available. Each of these can hold their own `background`, `foreground`, and `highlight` options.

Reviewer: Just checking if this one was missed or skipped? Since this is JS and we expect plugin developers to write JS I think we should use JS terminology as well - thus `option -> property`.

* appHeader
* paneHeader
* editor
* dialog
* transparentOverlay
* sidebar

> * link

Reviewer: These are all the styles available, appears there are a few more now https://github.com/Kong/insomnia/blob/cdf95f64d60d203ec0157b939f012b4ea80c754f/packages/insomnia-app/app/plugins/misc.ts#L44-L57

1. Create a new project for your theme. If you plan to make this theme available via NPM, start your project name with `insomnia-plugin-`. The following example is called `insomnia-plugin-dark-colorblind-theme`.

2. Following plugin instructions, write your [plugin package.json](https://docs.insomnia.rest/insomnia/introduction-to-plugins#plugin-packagejson). In your entry file, export your theme(s).

3. Start with a baseline template. Each section of the configuration has `background`, `foreground`, and `highlight` options, with modifiable color options.

  See the code comments for more information about which color option names correspond with which UI elements.

  {:.alert .alert-primary}

>   **Note**: The code comments below are not comprehensive and styles may apply elsewhere.

Reviewer: 👍🏽

```ts

> // If you want to generate your plugin from your entry file, 
> // use module.exports.themes = [{ ... }] instead. 

Reviewer: ```suggestion
// If you want to export your theme from the entry file, 
// use module.exports.themes = [{ ... }] instead. 
```

I do think it would be better for this not to be a comment, however. Like the other plugin examples, it should be possible for a user to copy-paste the entire example snippet directly into the `main.js` file that is created by the generate plugin button from inside the app. This is the least friction, as we don't assume a directory structure with the examples in docs

Reviewer: Thank you for updating this one 🙌

module.exports = {
  name:        'dark-colorblind', // theme name in kebab-case

>   displayName: 'Dark Colorblind', // formatted theme name

Reviewer: 👍🏽

  theme: {
    // Background, foreground, and highlight values nested directly in the theme 

>     // object will serve as the default overwrites for all options you add.

Reviewer: ```suggestion
    // object will serve as the default overwrites for all properties you add.
```

The following `style` options are available. Each of these can hold their own `background`, `foreground`, and `highlight` options.

* appHeader
* dialog
* dialogFooter
* dialogHeader
* dropdown
* editor
* link
* overlay
* pane
* paneHeader
* sidebar
* sidebarHeader
* sidebarList
* tooltip
* transparentOverlay

## Custom CSS

In addition to baseline edits and sub-components, you can add custom CSS using the `rawCss` option.

{:.alert .alert-primary}

> **Note**: Generally, opt to use predefined parameters to customize your theme rather than using custom CSS. This ensures that your theme will not break in the future as we make changes to Insomnia.

Reviewer: ```suggestion
**Note**: Generally, opt to use predefined properties to customize your theme rather than using custom CSS. This ensures that your theme will not break in the future as we make changes to Insomnia.
```
